Source/DafnyDriver/DafnyDriver.cs

```diff
@@ -662,28 +662,30 @@ public override void ReportBplError(IToken tok, string message, bool error, Text
       string targetProgramText;
       var otherFiles = new Dictionary<string, string>();
       {
-        var fileQueue = new Queue<FileTargetWriter>();
-        using (var wr = new TargetWriter(0)) {
-          compiler.Compile(dafnyProgram, wr);
-          var sw = new StringWriter();
-          wr.Collect(sw, fileQueue);
-          targetProgramText = sw.ToString();
-        }
-
-        while (fileQueue.Count > 0) {
-          var wr = fileQueue.Dequeue();
-          var sw = new StringWriter();
-          wr.Collect(sw, fileQueue);
-          otherFiles.Add(wr.Filename, sw.ToString());
+        var output = new TargetWriter();
+        compiler.Compile(dafnyProgram, output);
+        var writerOptions = new WriterState();
+        var targetProgramTextWriter = new StringWriter();
+        var files = new Queue<FileSyntax>();
+        output.Render(targetProgramTextWriter, 0, writerOptions, files);
+        targetProgramText = targetProgramTextWriter.ToString();
+
+        while(files.Count > 0)
+        {
+          var file = files.Dequeue();
+          var otherFileWriter = new StringWriter();
+          writerOptions.HasNewLine = false;
+          file.Tree.Render(otherFileWriter, 0, writerOptions, files);
+          otherFiles.Add(file.Filename, otherFileWriter.ToString());
         }
       }
       string callToMain = null;
-      if (hasMain) {
-        using (var wr = new TargetWriter(0)) {
-          string baseName = Path.GetFileNameWithoutExtension(dafnyProgramName);
-          compiler.EmitCallToMain(mainMethod, baseName, wr);
-          callToMain = wr.ToString(); // assume there aren't multiple files just to call main
-        }
+      if (hasMain)
+      {
+        var callToMainTree = new TargetWriter();
```

Comment:
Does this `TargetWriter` need to be closed? If so, the previous `using` statement would be useful here.

Reply:
> Does this TargetWriter need to be closed?

It doesn't. However, it currently still inherits from `TextWriter` which has an empty virtual `Dispose` method, so it would be cleaner to keep the `using`. I'll bring it back.

In a follow-up PR I'm planning on removing `TextWriter` as a base class, so then I'll take it out again ^^